infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentManager.java

```diff
@@ -0,0 +1,449 @@
+                .values()
+                .stream()
+                .sorted(Comparator.comparingLong(CompactionMetadata::getOrdinal).reversed())
+                .skip(logParams.protectedSegments)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Close the old segment and rename the new segment file to the old segment file.
+     * After this call, subsequent or pending calls to acquire a segment would return the new segment.
+     * NOTE: the caller needs to make sure that there are no concurrent writers by either locking
+     * or ensuring the segment is fully committed and cannot be written to.
+     */
+    void remapCompactedSegment(AbstractLogSegment oldSegment, AbstractLogSegment newSegment) {
+        long ordinal = newSegment.getOrdinal();
+        LoadingCache<Long, ? extends AbstractLogSegment> segmentCache;
+
+        if (newSegment instanceof StreamLogSegment) {
+            segmentCache = streamSegmentCache;
+        } else if (newSegment instanceof GarbageLogSegment) {
+            segmentCache = garbageSegmentCache;
+        } else {
+            throw new IllegalArgumentException("remapCompactedSegment: Unknown segment class: "
+                    + newSegment.getClass());
+        }
+
+        segmentCache.asMap().compute(ordinal, (ord, segment) -> {
+            if (segment != null) {
+                // Close the segment when it is not referenced. We don not need to force
+                // close it because writers cannot write anything new to the segment as it
+                // is consolidated before it could be compacted and readers are allowed to
+                // read from un-compacted segment.
+                segment.close(false);
+            }
+
+            Path oldPath = Paths.get(oldSegment.getFilePath());
+            Path newPath = Paths.get(newSegment.getFilePath());
+            try {
+                long oldFileSize = Files.size(oldPath);
+                // Rename and replace the old file with the new file.
+                Files.move(newPath, oldPath, StandardCopyOption.ATOMIC_MOVE,
+                        StandardCopyOption.REPLACE_EXISTING);
+                // Update log size quota to remove the space taken by the old file.
+                logSizeQuota.release(oldFileSize);
+
+            } catch (IOException e) {
+                log.error("remapCompactedSegment: exception when renaming file " +
+                        "{} to {}", newPath, oldPath, e);
+                throw new RuntimeException(e);
+            }
+
+            // Update the compactionMetadata map with the new CompactionMetadata.
+            segmentCompactionMetadata.put(ordinal, newSegment.getCompactionMetaData());
+
+            // Remove the segment mapping.
+            return null;
+        });
+    }
+
+    /**
+     * Close and remove the segment from the cache.
+     *
+     * @param ordinal ordinal of the segment to close
+     */
+    void close(long ordinal) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)